components/layout/layout_task.rs

```diff
@@ -98,26 +98,34 @@ const DISPLAY_PORT_THRESHOLD_SIZE_FACTOR: i32 = 4;
 ///
 /// This needs to be protected by a mutex so we can do fast RPCs.
 pub struct LayoutTaskData {
+    pub frozen: LayoutTaskFrozen,
+    pub unfrozen: LayoutTaskUnfrozen,
+}
+
+pub struct LayoutTaskFrozen {
+    /// The size of the viewport.
+    pub screen_size: Size2D<Au>,
+
+    /// Performs CSS selector matching and style resolution.
+    pub stylist: Stylist,
+
+    /// The position and size of the visible rect for each layer. We do not build display lists
+    /// for any areas more than `DISPLAY_PORT_SIZE_FACTOR` screens away from this area.
+    pub visible_rects: Arc<HashMap<LayerId, Rect<Au>, DefaultState<FnvHasher>>>,
+}
```

Comment:
Frozen probably isn't the best terminology, but I was having trouble thinking of something better.

When we do layout, we have to build a SharedLayoutContext.  To avoid extraneous locking/copying, it needs to have a pointer into the current LayoutTaskData.  (There's probably more unnecessary copying that could be avoided as a followup, but I wanted to keep this from getting too complicated.)  However, some bits need to be mutated while the SharedLayoutContext exists.  So this patch splits LayoutTaskData into the bits which are immutably borrowed by SharedLayoutContext and the bits which need to be mutated.  The "frozen" section is the one which is borrowed, and therefore can't be mutated.